File: 2025/ReminderApp/backend/src/main/java/com/reminder/service/ReminderService.java
package com.reminder.service;

import com.reminder.model.Reminder;
import com.reminder.repository.ReminderRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class ReminderService {

    @Autowired
    private ReminderRepository reminderRepository;

    public List<Reminder> getAllReminders() {
        return reminderRepository.findAll();
    }

    public Reminder getReminderById(Long id) {
        Optional<Reminder> reminder = reminderRepository.findById(id);
        return reminder.orElse(null);
    }

    public List<Reminder> searchReminders(String query) {
        return reminderRepository.searchByTitleOrDescription(query);
    }

    public List<Reminder> getRemindersByCategory(String category) {
        return reminderRepository.findByCategory(category);
    }

    public List<Reminder> getRemindersByStatus(Reminder.Status status) {
        return reminderRepository.findByStatus(status);
    }

    public List<Reminder> getRemindersByPriority(Reminder.Priority priority) {
        return reminderRepository.findByPriority(priority);
    }

    public List<Reminder> getUpcomingReminders() {
        return reminderRepository.findUpcomingReminders(LocalDateTime.now());
    }

    public List<Reminder> getOverdueReminders() {
        return reminderRepository.findOverdueReminders(LocalDateTime.now());
    }

    public Reminder createReminder(Reminder reminder) {
        return reminderRepository.save(reminder);
    }

    public Reminder updateReminder(Long id, Reminder reminderDetails) {
        Optional<Reminder> optionalReminder = reminderRepository.findById(id);
        if (optionalReminder.isPresent()) {
            Reminder reminder = optionalReminder.get();
            reminder.setTitle(reminderDetails.getTitle());
            reminder.setDescription(reminderDetails.getDescription());
            reminder.setReminderDate(reminderDetails.getReminderDate());
            reminder.setPriority(reminderDetails.getPriority());
            reminder.setStatus(reminderDetails.getStatus());
            reminder.setCategory(reminderDetails.getCategory());
            return reminderRepository.save(reminder);
        }
        return null;
    }

    public Reminder markAsCompleted(Long id) {
        Optional<Reminder> optionalReminder = reminderRepository.findById(id);
        if (optionalReminder.isPresent()) {
            Reminder reminder = optionalReminder.get();
            reminder.setStatus(Reminder.Status.COMPLETED);
            return reminderRepository.save(reminder);
        }
        return null;
    }

    public boolean deleteReminder(Long id) {
        if (reminderRepository.existsById(id)) {
            reminderRepository.deleteById(id);
            return true;
        }
        return false;
    }
}

File: 2025/ReminderApp/frontend/src/app/services/reminder.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Reminder } from '../models/reminder';

@Injectable({
  providedIn: 'root'
})
export class ReminderService {
  private apiUrl = 'http://localhost:8096/api/reminders';

  constructor(private http: HttpClient) {}

  getAllReminders(): Observable<Reminder[]> {
    return this.http.get<Reminder[]>(this.apiUrl);
  }

  getReminderById(id: number): Observable<Reminder> {
    return this.http.get<Reminder>(`${this.apiUrl}/${id}`);
  }

  searchReminders(query: string): Observable<Reminder[]> {
    const params = new HttpParams().set('query', query);
    return this.http.get<Reminder[]>(`${this.apiUrl}/search`, { params });
  }

  getRemindersByStatus(status: string): Observable<Reminder[]> {
    return this.http.get<Reminder[]>(`${this.apiUrl}/status/${status}`);
  }

  getRemindersByCategory(category: string): Observable<Reminder[]> {
    return this.http.get<Reminder[]>(`${this.apiUrl}/category/${category}`);
  }

  getRemindersByPriority(priority: string): Observable<Reminder[]> {
    return this.http.get<Reminder[]>(`${this.apiUrl}/priority/${priority}`);
  }

  getUpcomingReminders(): Observable<Reminder[]> {
    return this.http.get<Reminder[]>(`${this.apiUrl}/upcoming`);
  }

  getOverdueReminders(): Observable<Reminder[]> {
    return this.http.get<Reminder[]>(`${this.apiUrl}/overdue`);
  }

  createReminder(reminder: Reminder): Observable<Reminder> {
    return this.http.post<Reminder>(this.apiUrl, reminder);
  }

  updateReminder(id: number, reminder: Reminder): Observable<Reminder> {
    return this.http.put<Reminder>(`${this.apiUrl}/${id}`, reminder);
  }

  markAsCompleted(id: number): Observable<Reminder> {
    return this.http.put<Reminder>(`${this.apiUrl}/${id}/complete`, {});
  }

  deleteReminder(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}


